### PR TITLE
feat: allow rewriters to modify parent result

### PR DESCRIPTION
**Pull request by gregbty**
This updates the behavior of `Rewriter` so that the parent result is modified and returned instead of expecting the result of the modified key path. This is a follow-up to #20. I have successfully used this to create a rewriter capable of renaming fields.

I pushed this as a breaking change but not sure if this warrants another major release.

### CODE REVIEW

File: src/rewriters/NestFieldOutputsRewriter.ts [hunk 1]
```diff
@@ -76,10 +76,11 @@ class NestFieldOutputsRewriter extends Rewriter {
       ) {
         const rewrittenResponse = { ...pathResponse, ...pathResponse[this.newOutputName] };
         delete rewrittenResponse[this.newOutputName];
-        return rewrittenResponse;
+        response[key] = rewrittenResponse;
```
Comment (chanind) [inline src/rewriters/NestFieldOutputsRewriter.ts:79]: nit: it would be better to return a new object rather than modifying the input params directly.  Maybe return something like: `return { ...response, [key]: rewrittenResponse }` to avoid modifying the input.

File: src/rewriters/ScalarFieldToObjectFieldRewriter.ts [hunk 2]
```diff
@@ -49,14 +49,14 @@ class ScalarFieldToObjectFieldRewriter extends Rewriter {
...
 
   public rewriteResponse(response: any, key: string | number) {
-    const pathResponse = super.rewriteResponse(response, key);
-
     if (typeof response === 'object') {
+      const pathResponse = response[key];
+
       // undo the nesting in the response so it matches the original query
-      return pathResponse[this.objectFieldName];
+      response[key] = pathResponse[this.objectFieldName];
```
Comment (chanind) [inline src/rewriters/ScalarFieldToObjectFieldRewriter.ts:56]: same here, don't modify function inputs

**Review by chanind — CHANGES_REQUESTED** [review]
Apologies for the slow turn-around on this! Just 2 minor nits with modifying function inputs (really need to set up linting for this), but otherwise looks good!

**Comment by gregbty** [comment]
I resolved the formatting issues. I noticed that the array based rewrite test was failing, so I reverted the logic used for rewriting array items. This means that the feature is supported for only non-array parents. Is that fine, or would you rather not have that limitation.

**Comment by chanind** [comment]
As long as tests are passing it should be OK I think!

**Review by chanind — APPROVED** [review]
Looks good!

**Comment by Sytten** [comment]
This PR broke production for fields returning arrays, I am pushing a fix soon. This is quite critical @chanind @gregbty

**Comment by chanind** [comment]
Ack thanks for finding this! If you push a fix and a test we'll get this merged ASAP

**Comment by gregbty** [comment]
Sorry about that @Sytten, looks like there were a lack of tests for nested arrays which is why it was caught.

**Comment by Sytten** [comment]
This will all be resolved by pr #29, I added more tests but I still feel we need more to really ensure we dont break edge cases.

**Comment by chanind** [comment]
Sorry about that! I had naively assumed our tests were good enough to cover most cases, but clearly that was not true. Thanks so much for the quick fix for this @Sytten!
